### ISSUE TITLE
SharpMind.GPU: the inference kernels measure as a slowdown — replace with a training engine?

### ISSUE THREAD

**Issue by MBrekhof** [Integral2u/SharpMind#13]
I went to make `SharpMind.GPU` faster and concluded it needs replacing rather than fixing. That's a big enough call that I want to ask before proposing it, especially because it trades away part of a claim on the front page.

## What the current path does

Three findings, all verified by running the code rather than reading it:

1. **It never selects the GPU on a non-NVIDIA machine.** `GPUActivationKernels.cs:18` passes `!GPUMode.Cuda.Equals(BestBackend)` into `GetPreferredDevice`, whose parameter is `preferCPU`. Backend detection returns OpenCL on my machine, which isn't Cuda, so it passes `preferCPU: true`. I evaluated the expression: it returns `[CPU] CPUAccelerator`. ILGPU *does* see the GPU — `[OpenCL] gfx1152, 13.2 GB` — it's just never chosen.

2. **The kernels round-trip per output column.** `VecDotQ8_0_GPU` is called once per column, and each call does two `Marshal.Copy` host copies, four `Allocate1D` device allocations, a `LoadAutoGroupedStreamKernel`, a launch, a full `acc.Synchronize()` and a device-to-host readback, then sums 256 partials on the CPU. For the LM head that's 151,936 device round-trips per token.

3. **No F16 or F32 kernels exist** (documented at `GpuExtensions.cs:22-28`), so an fp16 model can't use a GPU matmul even in principle.

Net effect: ticking "Use GPU" makes things slower, silently. That part I'd argue is worth fixing either way.

## What I measured

Rented an RTX 3090 and benchmarked one Linear training step (3 GEMMs, device-resident) across backends — full harness and logs in the branch under `tools/GpuSpike/`:

| shape (M=256) | ILGPU tiled16 | cuBLAS SGEMM | cuBLAS TF32 |
|---|---|---|---|
| up_proj 896×4864 | 1.86 TF | 18.2 | 21.3 |
| lm_head 896×151936 | 1.83 TF | 19.2 | 31.0 |
| 7B up_proj 3584×18944 | 1.86 TF | 19.8 | 30.3 |

A hand-written ILGPU GEMM sits at ~5% of the card's FP32 peak — 10× behind cuBLAS. And the interesting bit: **ILGPU buffers + ILGPU stream + cuBLAS is identical to raw cuBLAS**, so the two compose at zero cost.

## What I built on that

`GpuBackpropEngine` — ILGPU keeps memory, streams and every custom kernel in pure C# (RMSNorm, RoPE, gated activation, causal GQA attention, fused cross-entropy, all forward and backward); cuBLAS does only the GEMMs, with the ILGPU tiled kernel as the fallback where there's no CUDA. It matches `BackpropEngine`'s loss and LoRA gradients to 1.44e-6 worst case across 24 parameters, runs through the real `TrainLoop` with accumulation/clipping/checkpoint/resume, and measures **553 ms/step against the CPU engine's 8,420 ms** on the same machine — 0.5B, LoRA r=8, batch 4 × seq 64.

## The cost, stated plainly

`README.md:21` says *"no dependency on llama.cpp, PyTorch, or any native runtime for its core path."* cuBLAS is a native dependency. The CPU path stays clean and the claim survives with "for its core path" doing the work — but anyone who opts into GPU training now needs the CUDA toolkit's cuBLAS. There's no way to have cuBLAS speed without cuBLAS; the alternative is writing GEMMs and accepting ~5% of peak, which for training is not a real option.

Two smaller costs: `ILGPU.Algorithms` becomes a dependency of `SharpMind.GPU` (the CUDA backend has no intrinsic for `Math.Exp` in a kernel — OpenCL hides this, so it only appears on real CUDA), and the `[PuzzlePeice]` GPU kernels go away, which costs `README.md:243` its worked example of cross-assembly kernel contribution. The mechanism is unchanged; it just loses its one live demonstrator.

## The questions

1. Is trading inference-GPU kernels for a training engine a direction you want at all? If not, I'd still suggest fixing or removing the current path — a switch that silently makes things slower is worse than no switch.
2. If yes: delete the old kernels, or keep them alongside? I deleted them, on the grounds that nothing selected them correctly anyway — but that's reversible if you'd rather they stayed.
3. Is the cuBLAS dependency acceptable for an opt-in training path, given the README claim?
4. GPU inference is worth having later — llama.cpp's Vulkan build gets 2.13× decode and 1.69× prefill over CPU on this same iGPU, so the ceiling is real. Would you want that built on this engine's infrastructure (device residency, arena, kernels) once training is settled?

Work is on `MBrekhof:feat/gpu-training` if it's useful to look at — 915 tests, spec and plan under `docs/superpowers/`. Happy to split it up, reshape it, or drop it depending on the answers.

**Comment by Integral2u** [comment]
_Never_ Selects GPU on non NVIDIA machine can fix _context.GetPreferredDevice(GPUMode.Cpu.Equals(GPUSharpMindConfig.BestBackend))
The other 2 The kernels round-trip per output column, No F16 or F32 kernels exist I can look at later but before that happens.  More important context, bar "Never Selects GPU on non NVIDIA" it's not really an issue in the sense the Sharpmind.GPU was only intended as a proof of concept for allowing 3rd parties to expand/replace and was done before some of the Plugin work was done.  I had a fleeting little hope it may show some improvement, but I didn't expect much from my poor little old system. not anything near RTX 3090, try 128mb integrated Radeon Vega GFX.  But you have nailed a pretty big nail on the head and proof that flexibility, customization is worth provisioning for.  What I'm thinking initially is you could build your version_(I would be happy to reference in the ReadMe and or bundle in CUI)_which  could work as is no additional support but I would defiantly support a GPU plugin interface that could be tied into CUI so users of a UI could choose which plugin(GPU/NPU/Other) they want to use and which types of [PuzzlePeice]'s they wish to utilize.

**Comment by Integral2u** [comment]
can you provide full link to MBrekhof:feat/gpu-training, 100% something i want to read in there but i can't find it.  yep i is old

**Comment by MBrekhof** [comment]
Here you go: https://github.com/MBrekhof/SharpMind/tree/feat/gpu-training

Diff view against your master: https://github.com/Integral2u/SharpMind/compare/master...MBrekhof:SharpMind:feat/gpu-training

Where to look: `SharpMind.GPU/` is the LoRA training engine (ILGPU — CUDA/OpenCL/CPU backends), and `tools/GpuSpike/` is the benchmark harness the numbers came from (includes the RunPod RTX 3090 runbook). Measured on a 3090: 553 ms/step, ~15x the CPU training engine, gradients matching CPU to 1.44e-6.

Heads up: I'll be rebasing the branch onto your latest master shortly, so commit hashes will change — the branch link stays valid. A GPU plugin interface in CUI sounds like the right shape; happy to take that discussion wherever you prefer.

**Comment by Integral2u** [comment]
Before I start, just impressive work.
Unclear why GpuExtensions was removed GPU, TPU, NPU what ever accelerator is to be opt in only.
I also cannot allow these type of actions in the main line.
[DllImport(Lib, EntryPoint = "cublasCreate_v2")] public static extern int Create(out IntPtr handle);
NativeResolver.  Yes I know there are benefits.  The use of ILGPU in SharpMind.GPU is because it's is Opt-in everything can function without the concept of a GPU, The ILGPU package is Dot Net code.

I have absolutely no issue with it being a stand-alone package. You would have full control of it and can update at your leisure. (No doubt your package will be faster)

The intention is to keep GPU specific or for that matter any add on variant out of the core packages. I don't consider CUI core but I do consider Training core.  We would need to devise an IAcceleratorPlugin which need to be exposed in SharpMind.Core so any provided DLL in a Plugins Folder or provided assembly can be boot strapped to provide what features is supports , overloads in mappings, we can extend mappings and create factories to product specific layer types if needed.  Do some ground work on the core now to support the advancements,

**Comment by MBrekhof** [comment]
Thanks — and agreed on all three points: accelerators stay opt-in, no P/Invoke in the main line, Training is core. The split you describe is the right one, and I'd like to do the ground work.

**What I'll open next, as its own issue/PR (small, pure .NET, no accelerator code):**

- `IAcceleratorPlugin` in `SharpMind.Core` — a plugin declares what it supports and hands out an `ITrainingEngine` (the seam that already exists: `ForwardBackward(batch)`, with `TrainLoop` keeping AdamW, clipping and checkpointing on the host). Discovery from the Plugins folder, same bootstrap as the existing tool plugins.
- Room for the layer-type factories / mapping overloads you mention, but I'd start engine-level because that's the seam with a working implementation behind it, and grow it from real need.

One question so I don't guess: do you want a **pure-ILGPU reference engine in-tree** as the opt-in default (restores what `GpuExtensions` offered, no native deps, ~9x the CPU engine by my numbers), or keep the tree accelerator-free and reference external plugins only? Either works for me.

**The cuBLAS variant stays out of your tree** — I'll ship it as a stand-alone plugin package (working name `SharpMind.GPU.Cuda`) that implements the same interface. A README mention would be great once the seam is in.

On `GpuExtensions`: I removed it because the inference kernels behind it measured as a slowdown on every backend I tried, and I didn't want to advertise a path that made things slower. Deleting was the wrong instrument — an opt-in seam is. The PR above puts that surface back in the shape you describe.

I'll hold the training branch as-is and won't touch the core until we've agreed the interface here.

**Comment by Integral2u** [comment]
"pure-ILGPU reference" once the IAcceleratorPlugin is defined SharpMind will be needing to implement.  Its only really referenced in tests and CUI.  Tests is fine people can also add via code not needing the plugin path but CUI will need to be updated so when setting options any available plugin can be selected for accelerator options.  So I'll be un-referencing it leaving it in the main tree but having build export to the CUI Plugins folder.  I can and special mention of your version and we can work out a way to also include the cuBLAS variant  dll"s into the packaged CUI.  Hmm probably wise to allow folders for Plugins and CUI can scan reclusively so we don't get stuck in a dll hell. The current ILGPU version on the main tree will have no special treatment and will be bound by same limitations so nailing IAcceleratorPlugin is a priority for all.

**Comment by MBrekhof** [comment]
Thanks — that settles it. Before I open the ground-work PR, here is the seam I'd like to build, so we agree on shape before there's a diff to argue about.

**One host interface in `SharpMind.Core`, capabilities by type:**

```csharp
// SharpMind.Core/Plugins/IAcceleratorPlugin.cs
public interface IAcceleratorPlugin
{
    string Name { get; }            // stable id: "cuda", "ilgpu" — shown in CUI, stored in the .smmt job
    string Description { get; }
    IReadOnlyList<object> Capabilities { get; }   // host picks by type
}

// SharpMind.Core/Plugins/IMappingOverrides.cs — kernel level, what SharpMind.GPU does today
public interface IMappingOverrides
{
    IReadOnlyDictionary<string, string> GetOverrides(SharpMindConfig config);
}
```

**Training capability in `SharpMind.Training`** (Core can't name `Transformer`/`TrainingBatch`):

```csharp
public interface ITrainingEngine : IDisposable
{
    /// Forward + backward for one batch, accumulates into Parameter.Grad, returns the mean loss.
    /// Caller owns zeroing, clipping and the optimizer step.
    float ForwardBackward(TrainingBatch batch, CancellationToken ct = default);
}

public interface ITrainingEngineFactory
{
    /// Null when this plugin can't train this model/config; reason is shown in the CUI.
    ITrainingEngine? TryCreate(TrainingEngineContext ctx, out string? reason);
}

public sealed record TrainingEngineContext(
    Transformer Model, IReadOnlyList<Parameter> Parameters, GradientMapping Mapping,
    SharpMindConfig Config, ILoss<int> Loss, int BatchSize, int SeqLen,
    int IgnoreId = -100, float LabelSmoothing = 0f);
```

Why two levels rather than mapping overrides alone: a GPU only pays for training when the whole step stays resident (weights, activations, loss on device) — per-kernel overrides round-trip through host memory on every op, which is the slowdown this issue measured. Inference can stay mapping-level, so the existing `SharpMind.GPU` fits the same plugin interface unchanged.

**Wiring:** `TrainLoop` gets an optional `ITrainingEngine` (default = a `CpuTrainingEngine` wrapping today's `BackpropEngine`, bit-identical); `AcceleratorLoader.LoadFrom(folder)` next to `PluginAssemblyResolver` scans recursively; `TrainJobSettings.Accelerator` (null = CPU) + one radio group in the training wizard. Pure .NET, no accelerator code, ~350 lines with tests. I'd leave making `PluginLoader`/`ComponentRegistry` recursive to a follow-up unless you want it in the same PR.

Three questions before I cut the branch: (1) are you fine with the engine contracts living in `SharpMind.Training` rather than Core, given the type dependencies? (2) do you want the CUI accelerator option in this PR, or would you rather wire that yourself alongside un-referencing `SharpMind.GPU`? (3) if you'd prefer the engine to be *chosen* through the mapping like everything else — a `KeyTrainingEngine` entry resolved by a `MappingHash`-cached factory, with `TrainJobSettings.Accelerator` just writing that override — I'm happy to build it that way; the engine body stays opaque either way, only the selection idiom changes.

If you'd rather keep training strictly per-kernel I'll understand — the fork can carry the optional `engine` parameter on `TrainLoop` and ship the CUDA engine as a plugin against that; I'd just rather the seam lived upstream so anyone's accelerator can use it.

**Comment by Integral2u** [comment]
(1) Yes I'm fine with that.
(2) "do you want the CUI accelerator option in this PR" yes - I figured unwiring SharpMind.GPU would have to happen but no point me touching that space as it could have conflicted with what you need to end up doing to make it work.  Figured would be to many fingers in the pie.
(3) I would have thought this and 2 are linked, proceed to something like .WithGPU(....options) that maps method name variant or provides available interfaces.  I know the basic SharpMind.GPU will need changes but best to bite the bullet now.  For simplicity I don't think we need mixing of accelerator projects but undoubtably somebody may end up wanting to replace one mapping for experimentation.  WithGPU is just a pattern to make the builder simple, really not an interface.

If you'd rather keep training strictly per-kernel I'll understand -no it doesn't have to be, it's apparent the work you have done shows considerable improvement so no point of hamstringing that.

**Comment by Integral2u** [comment]
Resolved with PR #28
